git_patrol.py

```diff
@@ -159,30 +162,38 @@ def __init__(self):
   # Provide a few default substitutions that Google Cloud Build would fill in
   # if it was launching a triggered workflow. See link for details...
   # https://cloud.google.com/cloud-build/docs/configuring-builds/substitute-variable-values
-  arg_substitutions = '--substitutions='
+  substitutions_list = []
   if git_ref.startswith('refs/tags/'):
-    arg_substitutions += 'TAG_NAME={},'.format(
-        git_ref.replace('refs/tags/', ''))
+    substitutions_list.append(
+        'TAG_NAME={}'.format(git_ref.replace('refs/tags/', '')))
   elif git_ref.startswith('refs/heads/'):
-    arg_substitutions += 'BRANCH_NAME={},'.format(
-        git_ref.replace('refs/heads/', ''))
+    substitutions_list.append(
+        'BRANCH_NAME={}'.format(git_ref.replace('refs/heads/', '')))
 
-  # Generate a substitution string from the target config.
-  # TODO(brian): Handle no substitutions.
-  subs_config = (
-      ','.join(
-          '{!s}={!s}'.format(
-              k, v) for (k, v) in config['substitutions'].items()))
-  arg_substitutions += subs_config
+  # Generate substitution strings from the target config.
+  if 'substitutions' in config:
+    substitutions_list += [
+        '{!s}={!s}'.format(k, v) for (k, v) in config['substitutions'].items()]
 
-  arg_sources = os.path.join(config_path, config['sources'])
+  # Populate the substitutions argument if needed.
+  arg_substitutions = ''
+  if substitutions_list:
+    arg_substitutions = '--substitutions=' + ','.join(substitutions_list)
+
+  # Support an optional source archive passed to the workflow.
+  arg_sources = '--no-source'
+  if 'sources' in config:
+    arg_sources = os.path.join(config_path, config['sources'])
 
   gcloud_subproc = await commands.gcloud(
       'builds', 'submit', '--async', arg_config, arg_substitutions, arg_sources)
-  stdout_bytes, _ = await gcloud_subproc.communicate()
+  stdout_bytes, stderr_bytes = await gcloud_subproc.communicate()
   returncode = await gcloud_subproc.wait()
   if returncode:
     logger.warning('gcloud builds submit returned %d', returncode)
+    logger.warning(
+        '\nstdout\n%s\nstderr\n%s', stdout_bytes.decode('utf-8', 'ignore'),
+        stderr_bytes.decode('utf-8', 'ignore'))
```

Comment:
this is repetitive from above...maybe add helper function

Reply:
Done.